builtin/providers/aws/resource_aws_instance.go

```diff
@@ -514,44 +530,130 @@ func resourceAwsInstanceCreate(d *schema.ResourceData, meta interface{}) error {
-			// IAM profiles can take ~10 seconds to propagate in AWS:
-			//  http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#launch-instance-with-role-console
-			if awsErr.Code() == "InvalidParameterValue" && strings.Contains(awsErr.Message(), "Invalid IAM Instance Profile") {
-				log.Printf("[DEBUG] Invalid IAM Instance Profile referenced, retrying...")
-				time.Sleep(2 * time.Second)
-				continue
-			}
+	instanceID := ""
+	onDemand := true
+
+	if sp := d.Get("spot_price").(string); sp != "" {
+		// FIXME: check that instance type is not t2.micro, t2.small, or t2.medium
+
+		onDemand = false
+		// FIXME: validate spot price casts to a sensible decimal
+
+		spotPlacement := &ec2.SpotPlacement{
+			AvailabilityZone: aws.String(d.Get("availability_zone").(string)),
+			GroupName:        aws.String(d.Get("placement_group").(string)),
+		}
+
+		launchSpec := &ec2.RequestSpotLaunchSpecification{
+			Placement:           spotPlacement,
+			BlockDeviceMappings: runOpts.BlockDeviceMappings,
+			ImageID:             runOpts.ImageID,
+			InstanceType:        runOpts.InstanceType,
+			UserData:            runOpts.UserData,
+			EBSOptimized:        runOpts.EBSOptimized,
+			IAMInstanceProfile:  runOpts.IAMInstanceProfile,
+			NetworkInterfaces:   runOpts.NetworkInterfaces,
+		}
+
+		if runOpts.SubnetID != nil {
+			launchSpec.SubnetID = runOpts.SubnetID
+		}
+
+		if runOpts.SecurityGroupIDs != nil {
+			launchSpec.SecurityGroupIDs = runOpts.SecurityGroupIDs
+		}
+
+		if runOpts.SecurityGroups != nil {
+			launchSpec.SecurityGroups = runOpts.SecurityGroups
+		}
+
+		if runOpts.KeyName != nil {
+			launchSpec.KeyName = runOpts.KeyName
+		}
+
+		// Build the spot instance struct
+		spotOpts := &ec2.RequestSpotInstancesInput{
+			SpotPrice:           aws.String(sp), // Required
+			InstanceCount:       aws.Long(1),
+			LaunchSpecification: launchSpec,
+		}
+
+		// Make the spot instance request
+		var spotResp *ec2.RequestSpotInstancesOutput
+		spotResp, err := conn.RequestSpotInstances(spotOpts)
+
+		request_id := *spotResp.SpotInstanceRequests[0].SpotInstanceRequestID
+		d.Set("spot_request_id", request_id)
+
+		spotStateConf := &resource.StateChangeConf{
+			// http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-bid-status.html
+			Pending:    []string{"start", "pending-evaluation", "pending-fulfillment"},
+			Target:     "fulfilled",
+			Refresh:    SpotInstanceStateRefreshFunc(conn, request_id),
+			Timeout:    10 * time.Minute,
+			Delay:      10 * time.Second,
+			MinTimeout: 3 * time.Second,
+		}
+
+		log.Printf("[DEBUG] waiting for spot bid to resolve... this may take several minutes.")
+		spotRequestRaw, err := spotStateConf.WaitForState()
+		spotRequest := spotRequestRaw.(*ec2.SpotInstanceRequest)
+
+		fallback := d.Get("spot_fallback").(bool)
+
+		if err != nil && !fallback {
+			return fmt.Errorf("Error while waiting for spot request (%s) to resolve: %s", request_id, err)
+		} else if fallback {
```

Comment:
> Here you fall back to a standard instance if the spot bid fails, but spot instances may succeed now and be pre-empted later.

I hadn't considered that edge case when writing this, thanks. This will only be the case if the request actually times out. The logic i had considered when writing was that the spot instance fails because the price does not resolve (bid too low) or amazon otherwise cannot provision the capacity. The intention of 'spot fallback' is to give a pricier on demand instance if the spot request cannot be fulfilled, aka 'try to get the cheap instance, but settle for full price'.

> I wonder if it would make more sense to have spot be a separate type of resource. Any thoughts about this?

I'm ambivalent on this too. On the one hand, having it as a separate resource feels cleaner. On the other hand, it's not very DRY because spot instances and on demand instances are almost identical - the only difference is how they are launched. Spot instances have the extra 'bid resolution' phase, but aside from that most of the parameters are the same. Making them a separate resource would basically entail copying this file and maintaining all this logic in two places.

Are spot instances really different than normal AWS instances? No, they aren't - they're identical instances. The only difference is the provisioning strategy. Spot instances take longer to provision, and provisioning is less certain. The upside though, is that you can typically save a **substantial** amount of money by using them, and they fit a lot of use cases (particularly for ephemeral compute resources) very nicely. So, if you're willing to wait, it can be worth it.